### ISSUE TITLE
module.exports proxyMiddelware throw Error

### ISSUE THREAD

**Issue by monstromatic** [Swiip/generator-gulp-angular#442]
Hello,

I just update your great generator but my usually settings throw now an error:
/project/gulpfile.js:24
  require('./gulp/' + file)(options);
                           ^
TypeError: object is not a function
    at /project/gulpfile.js:24:28
    at Array.map (native)
    at Object.<anonymous> (/project/gulpfile.js:23:4)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Liftoff.handleArguments (/usr/local/share/npm/lib/node_modules/gulp/bin/gulp.js:108:3)

In gulp/proxy.js, module.exports = [proxyMiddleware];

**Comment by monstromatic** [comment]
I tried 
module.exports = function(){
  return [proxyMiddleware];
};

No more error in terminal but the proxy doesn't work anymore...

**Comment by zckrs** [comment]
Unreadable issue see https://github.com/Swiip/generator-gulp-angular/blob/master/CONTRIBUTING.md#issues

**Comment by paweln1986** [comment]
I have got same issue. Below chages in **gulpfile.js** helps me and now proxy works as before update.  
```javascript
if(file==='proxy.js'){
   require('./gulp/' + file)
}else{
  require('./gulp/' + file)(options);
}
```

**Comment by jahumes** [comment]
I am also getting this issue. I have tried making changes to the target, but with no success. The following code causes ````gulp serve```` to fail. Below is my entire proxy file with just the change made to activate it.

````javascript
 /*jshint unused:false */

/***************

  This file allow to configure a proxy system plugged into BrowserSync
  in order to redirect backend requests while still serving and watching
  files from the web project

  IMPORTANT: The proxy is disabled by default.

  If you want to enable it, watch at the configuration options and finally
  change the `module.exports` at the end of the file

***************/

'use strict';

var httpProxy = require('http-proxy');
var chalk = require('chalk');

/*
 * Location of your backend server
 */
var proxyTarget = 'http://server/context/';

var proxy = httpProxy.createProxyServer({
  target: proxyTarget
});

proxy.on('error', function(error, req, res) {
  res.writeHead(500, {
    'Content-Type': 'text/plain'
  });

  console.error(chalk.red('[Proxy]'), error);
});

/*
 * The proxy middleware is an Express middleware added to BrowserSync to
 * handle backend request and proxy them to your backend.
 */
function proxyMiddleware(req, res, next) {
  /*
   * This test is the switch of each request to determine if the request is
   * for a static file to be handled by BrowserSync or a backend request to proxy.
   *
   * The existing test is a standard check on the files extensions but it may fail
   * for your needs. If you can, you could also check on a context in the url which
   * may be more reliable but can't be generic.
   */
  if (/\.(html|css|js|png|jpg|jpeg|gif|ico|xml|rss|txt|eot|svg|ttf|woff|woff2|cur)(\?((r|v|rel|rev)=[\-\.\w]*)?)?$/.test(req.url)) {
    next();
  } else {
    proxy.web(req, res);
  }
}

/*
 * This is where you activate or not your proxy.
 *
 * The first line activate if and the second one ignored it
 */

module.exports = [proxyMiddleware];
//module.exports = function() {
//  return [];
//};
````

**Comment by paweln1986** [comment]
Try to change gulpfile.js to 
```javascript
'use strict';

var gulp = require('gulp');
var gutil = require('gulp-util');
var _ = require('lodash');
var wrench = require('wrench');

var options = {
  src: 'src',
  dist: 'dist',
  tmp: '.tmp',
  e2e: 'e2e',
  errorHandler: function(title) {
    return function(err) {
      gutil.log(gutil.colors.red('[' + title + ']'), err.toString());
      this.emit('end');
    };
  }
};

wrench.readdirSyncRecursive('./gulp').filter(function(file) {
  return (/\.(js|coffee)$/i).test(file);
}).map(function(file) {
  if(file==='proxy.js'){
    require('./gulp/' + file)
  }
  else{
    require('./gulp/' + file)(options);
  }
});

gulp.task('default', ['clean'], function () {
    gulp.start('build');
});
```
It is ugly workaround but for me work.

**Comment by drmikeh** [comment]
I fixed the problem by changing 2 lines in gulp/server.js. I changes this:
```javascript
    if(middleware.length > 0) {
      server.middleware = middleware;
```
to this:
```javascript
    if(middleware().length > 0) {
      server.middleware = middleware();
````
and my proxy.js returns this:
```javascript
module.exports = function() {
  return [proxyMiddleware];
};
```

**Comment by kostonstyle** [comment]
Thanks

**Comment by andyperlitch** [comment]
This and other recent issues have occurred due to #335. I am about to issue a pull request to fix this particular problem.

**Comment by ronaldallanmojica** [comment]
When I applied your fix, it now serves the proxy but not the src

**Comment by hutufeia2009** [comment]
@ronaldallanmojica Can you tell me how to fix?

**Comment by shawn-simon** [comment]
I figured out the issue. ./proxy.js is getting required at the start and expects an options param. However in it's re-required later in server.js with out the function being executed. The commented out line wouldn't work. It should not be included in the initial require as @paweln1986  pointed out.

**Comment by zckrs** [comment]
Can you try if issue exist on master ?

**Comment by trob67** [comment]
Hi, I tried this today and got the same problems described above. I tried above suggestions but could not get it working. On load the whole page seem to try and redirect to whatever was in proxyTarget.

**Comment by Tallyb** [comment]
same issue. I bypassed it as following: 

module.exports = function() {
    return [proxyMiddleware]; // enable this line if you want a proxy
//  return []; //enable this line if you do not want a proxy
};

and then in server.js: 
var middleware = require('./proxy')(); //added a function call (namely paranthesis).

**Comment by trob67** [comment]
Still doesn't work for me. If I try TallyB change I get 
404 Not Found: Requested route ('localhost') does not exist.
as soon as I do gulp serve.

**Comment by shawn-simon** [comment]
Not sure if this is helpful but I actually rewrote that part locally. I remove the proxy / server JSON from the browsersync init, and removed proxy.js. Then I have gulp serve kick off nodemon for my express app, passing in configuration that tells it to inject the browserSync javascript snippet on the index. It works a lot better. The downside is that you need to get gulp working with a different current working directory. Now that it's working, I wouldn't go back to the other method, since you can't watch server files, or use serverside templating.

**Comment by Tallyb** [comment]
Care to share your code?

**Comment by trob67** [comment]
I do:
yo gulp-angular

make the change you suggest:
proxy.js 
var proxyTarget = 'http://rest-service.guides.spring.io/';

module.exports = function() {
	return [proxyMiddleware]; 
};

server.js
var middleware = require('./proxy')();

then:
gulp serve 

result displayed on browser at start up
404 Not Found: Requested route ('localhost') does not exist.

**Comment by ronaldallanmojica** [comment]
here is mine.. i removed all configuration of proxy.js and modified server.js. I installed 'http-proxy' and  added environment feature for dev, staging and prod :). Hope it helps.

    'use strict';

    var gulp = require('gulp');
    var browserSync = require('browser-sync');
    var browserSyncSpa = require('browser-sync-spa');
    var gutil = require('gulp-util');

    var util = require('util');

    var middleware = require('./proxy');
    var httpProxy = require('http-proxy');

    module.exports = function(options) {

      function browserSyncInit(baseDir, browser, env) {
        browser = browser === undefined ? 'default' : browser;
        env = env === undefined ? 'makdev' : env;
        var routes = null;
        if(baseDir === options.src || (util.isArray(baseDir) && baseDir.indexOf(options.src) !== -1)) {
          routes = {
            '/bower_components': 'bower_components'
          };
        }


        gutil.log('Configured for ' + gutil.colors.green('[' + env + '] environment'));

        var proxyStripped = httpProxy.createProxyServer({
            target: 'http://' + env + '.yourapi.com'
          });

        function fnProxyMiddleWare (req, res, next) {
          if (req.url.indexOf('/api/services') === 0) {
            proxyStripped.web(req, res);
          }
          else {
            next();
          }
        }

        var server = {
          baseDir: baseDir,
          routes: routes,
          middleware: [fnProxyMiddleWare]
        };

        if(middleware.length > 0) {
          server.middleware = middleware;
        }

        browserSync.instance = browserSync.init({
          startPath: '/',
          server: server,
          browser: browser
        });
      }

      browserSync.use(browserSyncSpa({
        selector: '[ng-app]'// Only needed for angular apps
      }));

      gulp.task('serve', ['watch'], function () {
        browserSyncInit([options.tmp + '/serve', options.src]);
      });

      gulp.task('serve:dev', ['build'], function () {
        browserSyncInit(options.tmp, [], 'dev' );
      });

      gulp.task('serve:staging', ['build'], function () {
        browserSyncInit(options.tmp, [], 'staging');
      });

      gulp.task('serve:test', ['build'], function () {
        browserSyncInit(options.tmp, [], 'test');
      });

      gulp.task('serve:prod', ['build'], function () {
        browserSyncInit(options.dist);
      });

      gulp.task('serve:e2e', ['inject'], function () {
        browserSyncInit([options.tmp + '/serve', options.src], []);
      });

      gulp.task('serve:e2e-dist', ['build'], function () {
        browserSyncInit(options.dist, []);
      });
    };

**Comment by shawn-simon** [comment]
Here's mine using nodemon instead of the proxy / watch server:

```
'use strict';

var gulp = require('gulp');
var browserSync = require('browser-sync');
var browserSyncSpa = require('browser-sync-spa');
var util = require('util');
var chalk = require('chalk');
var nodemon = require('gulp-nodemon');

module.exports = function(options) {

  function nodemonInit(build_type) {
    var configuration = {
      webclient: {
        browserSync: true,
        development: build_type === 'dist' ? false : true
      }
    }
    nodemon({
      script: 'server.js',
      env: {
        'NODE_CONFIG': JSON.stringify(configuration)
      }
    })
    console.log(chalk.blue("Mobcrush web server starting"))
  }

  function browserSyncInit(baseDir, browser) {
    browser = browser === undefined ? 'default' : browser;
    browserSync.instance = browserSync.init({
      startPath: '/',
      browser: browser,
      notify: false,
      ui: false,
      logSnippet: false,
    });
  }
  browserSync.use(browserSyncSpa({
    selector: '[ng-app]'// Only needed for angular apps
  }));

  gulp.task('serve', ['watch'], function () {
    browserSyncInit([options.tmp + '/serve', options.src]);
    nodemonInit()
  });

  gulp.task('serve:dist', ['build'], function () {
    browserSyncInit(options.dist);
    nodemonInit('dist')
  });

};
```

**Comment by zckrs** [comment]
New proxy https://github.com/Swiip/generator-gulp-angular/pull/593 
:tada: Just released v0.12.0 :tada:
Your issue should be fixed. I close but if you can retry your case with a new fresh install of generator you are welcome. Any feedback would be appreciated (positive comments too ^_^).
